File: csharp/examples/GtpChat.cs
// Two-party GTP (text) chat — minimal in-memory end-to-end example.
//
// Demonstrates:
//   - MLS two-party handshake
//   - GBP node bootstrap
//   - Sending text messages with CBOR and FlatBuffers codecs
//   - Idempotency: duplicate (sender, messageId) returns status "duplicate"
//
// Run from repo root:
//   dotnet script csharp/examples/GtpChat.cs
// Or add to a console project and run with `dotnet run`.

using GBPStack;

// --- MLS handshake -----------------------------------------------------------
using var aliceMls = MlsContext.Create("alice");
using var bobMls   = MlsContext.Create("bob");

bobMls.AcceptWelcome(aliceMls.Invite(bobMls.ExportKeyPackage()));
Console.WriteLine($"MLS epoch after invite: alice={aliceMls.Epoch}  bob={bobMls.Epoch}");

// --- GBP nodes ---------------------------------------------------------------
var gid = aliceMls.GroupId;
using var alice = GroupNode.Create(memberId: 1, gid);
using var bob   = GroupNode.Create(memberId: 2, gid);
alice.BootstrapAsCreator(aliceMls.Epoch);
bob.BootstrapAsJoiner(bobMls.Epoch);

// --- GTP clients -------------------------------------------------------------
using var gtpAlice = GtpClient.Create();
using var gtpBob   = GtpClient.Create();

// Send "hello" with default CBOR codec.
var frame = gtpAlice.Send(alice, aliceMls, target: 2, messageId: 1, "hello");
foreach (var ev in bob.OnWire(bobMls, frame.Wire))
{
    if (ev.Kind == "payload_received" && ev.StreamType == StreamType.Text)
    {
        var r = gtpBob.Accept(ev.Plaintext!, bobMls.Epoch, ev.Codec ?? PayloadCodec.Cbor);
        Console.WriteLine($"new (cbor):   text={r.Text}  status={r.Status}");
    }
}

// Send with FlatBuffers codec.
var frame2 = gtpAlice.Send(alice, aliceMls, target: 2, messageId: 2, "hello flatbuffers",
    codec: PayloadCodec.FlatBuffers);
foreach (var ev in bob.OnWire(bobMls, frame2.Wire))
{
    if (ev.Kind == "payload_received" && ev.StreamType == StreamType.Text)
    {
        var r = gtpBob.Accept(ev.Plaintext!, bobMls.Epoch, ev.Codec ?? PayloadCodec.Cbor);
        Console.WriteLine($"new (fbs):    text={r.Text}  codec={ev.Codec}");
    }
}

// Replay: same messageId=1 must come back as "duplicate".
var dup = gtpAlice.Send(alice, aliceMls, target: 2, messageId: 1, "hello");
foreach (var ev in bob.OnWire(bobMls, dup.Wire))
{
    if (ev.Kind == "payload_received" && ev.StreamType == StreamType.Text)
    {
        var r = gtpBob.Accept(ev.Plaintext!, bobMls.Epoch);
        Console.WriteLine($"replay:       status={r.Status}");  // → duplicate
    }
}

File: python/examples/gtp_chat.py
"""Two-party GTP (text) chat — minimal in-memory end-to-end example.

Demonstrates:
  - MLS two-party handshake
  - GBP node bootstrap
  - Sending text messages with CBOR and FlatBuffers codecs
  - Idempotency: duplicate (sender, message_id) returns status "duplicate"

Run:
    python python/examples/gtp_chat.py
"""

from gbp_stack import MlsContext, GroupNode, GtpClient, PayloadCodec

# --- MLS handshake -----------------------------------------------------------
with MlsContext.create("alice") as alice_mls, \
     MlsContext.create("bob")   as bob_mls:

    bob_kp = bob_mls.export_key_package()
    bob_mls.accept_welcome(alice_mls.invite(bob_kp))
    print(f"MLS epoch: alice={alice_mls.epoch}  bob={bob_mls.epoch}")

    # --- GBP nodes -----------------------------------------------------------
    gid = alice_mls.group_id
    with GroupNode.create(member_id=1, group_id=gid) as alice, \
         GroupNode.create(member_id=2, group_id=gid) as bob, \
         GtpClient.create() as gtp_alice, \
         GtpClient.create() as gtp_bob:

        alice.bootstrap_as_creator(alice_mls.epoch)
        bob.bootstrap_as_joiner(bob_mls.epoch)

        # Send "hello" with default CBOR codec.
        frame = gtp_alice.send(alice, alice_mls, target=2, message_id=1, text="hello")
        for ev in bob.on_wire(bob_mls, frame.wire):
            if ev.kind == "payload_received" and ev.stream_type == 2:
                r = gtp_bob.accept(ev.plaintext or b"", bob_mls.epoch,
                                   codec=ev.codec or PayloadCodec.CBOR)
                print(f"new (cbor):   text={r.text!r}  status={r.status}")

        # Send with FlatBuffers codec.
        frame2 = gtp_alice.send(alice, alice_mls, target=2, message_id=2,
                                text="hello flatbuffers",
                                codec=PayloadCodec.FLATBUFFERS)
        for ev in bob.on_wire(bob_mls, frame2.wire):
            if ev.kind == "payload_received" and ev.stream_type == 2:
                r = gtp_bob.accept(ev.plaintext or b"", bob_mls.epoch,
                                   codec=ev.codec or PayloadCodec.CBOR)
                print(f"new (fbs):    text={r.text!r}  codec={ev.codec}")

        # Replay: same message_id=1 must come back as "duplicate".
        dup = gtp_alice.send(alice, alice_mls, target=2, message_id=1, text="hello")
        for ev in bob.on_wire(bob_mls, dup.wire):
            if ev.kind == "payload_received" and ev.stream_type == 2:
                r = gtp_bob.accept(ev.plaintext, bob_mls.epoch)
                print(f"replay:       status={r.status}")  # → duplicate
